src/main/java/org/sagebionetworks/web/client/widget/ReactComponent.java

```diff
@@ -0,0 +1,225 @@
+        break;
+      }
+    }
+    return allChildrenAreReactComponents;
+  }
+
+  boolean isRenderedAsReactComponentChild() {
+    return (
+      getParent() instanceof ReactComponent &&
+      ((ReactComponent<?>) getParent()).allChildrenAreReactComponents()
+    );
+  }
+
+  /**
+   * This method
+   * returns the root ReactComponent that should be re-rendered.
+   *
+   * @return
+   */
+  ReactComponent<?> getRootReactComponentWidget() {
+    if (isRenderedAsReactComponentChild()) {
+      return ((ReactComponent<?>) getParent()).getRootReactComponentWidget();
+    } else {
+      return this;
+    }
+  }
+
+  @Override
+  public HandlerRegistration addClickHandler(ClickHandler handler) {
+    return addDomHandler(handler, ClickEvent.getType());
+  }
+
+  private void maybeCreateRoot() {
+    if (root == null && !isRenderedAsReactComponentChild()) {
+      root = ReactDOM.createRoot(this.getElement());
+    }
+  }
+
+  /**
+   * Override the current props of the React component.
+   * Because re-rendering the component will use `React.cloneElement`, old props must be explicitly set to `undefined`
+   * to remove them.
+   */
+  public void overrideProps(T props) {
+    this.props = props;
+    this.rerender();
+  }
+
+  public void render(ReactNode<?> reactNode) {
+    this.reactElement = reactNode;
+    maybeCreateRoot();
+
+    if (this.allChildrenAreReactComponents()) { // If all widget children are ReactNodes, clone the React component and add them as children
+      List<ReactComponent<?>> childWidgets = new ArrayList<>();
+      getChildren().forEach(w -> childWidgets.add(((ReactComponent<?>) w)));
+
+      ReactNode<?>[] childReactElements = childWidgets
+        .stream()
+        .map(ReactComponent::getReactElement)
+        .toArray(ReactNode<?>[]::new);
+
+      this.reactElement =
+        React.cloneElement(reactElement, this.props, childReactElements);
+    } else if (getChildren().size() > 0) {
+      // Create a callback ref that will allow us to inject the GWT children into the DOM
+      ReactComponentProps.CallbackRef refCallback = (Element node) -> {
+        if (node != null) {
+          // Once the DOM node is defined, inject each child
+          getChildren().forEach(w -> node.appendChild(w.getElement()));
+        }
+      };
+
+      if (this.props == null) {
+        this.props = (T) JsPropertyMap.of();
+      }
+      this.props.ref = refCallback;
+
+      this.reactElement =
+        React.cloneElement(
+          reactElement,
+          // Override the ref
+          this.props
+        );
+    }
+
+    // This component may be a React child of another component. If so, re-render the entire tree
+    // This can resolve issues where e.g. a component was set change visibility
+    ReactComponent<?> componentToRender = getRootReactComponentWidget();
+    if (componentToRender == this) {
+      root.render(this.reactElement);
+    } else {
+      componentToRender.rerender();
+    }
```

Comment:
Many bugs went away when I started re-rendering from the root node in cases where we had nested React components.